text/0000-repr-type-aliases.md

```diff
@@ -0,0 +1,84 @@
+- Feature Name: `repr_type_aliases`
+- Start Date: 2024-06-14
+- RFC PR: [rust-lang/rfcs#0000](https://github.com/rust-lang/rfcs/pull/0000)
+- Rust Issue: [rust-lang/rust#0000](https://github.com/rust-lang/rust/issues/0000)
+
+# Summary
+[summary]: #summary
+
+Primitive representations on enums now accept type aliases, meaning that in addition to primitives like `#[repr(u32)]`, `#[repr(type = core::ffi::c_int)]` and `#[repr(type = my_type)]` are now accepted.
+
+# Motivation
+[motivation]: #motivation
+
+For the same reasons why type aliases are useful, having type aliases in `repr` attributes would also be useful. A few examples:
+
+* Types depend on an external API whose exact size may be uncertain. (e.g. `core::ffi::c_int`, `gl::types::GLsizei`)
+* An internal API might want to be able to easily change a type later.
+* The intent behind a type alias may be clearer than simply using the primitive directly.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+Enums allow `#[repr(type = ...)]` attributes to offer an explicit discriminant type. (`...` can be any primitive integer type, like `u8`, `i32`, or `usize`, but not `char`.) If all variants of the enum are unit variants, this means that the enum will be easily castable to `type` using `as`. Otherwise, the discriminant will still be of the specified type, but unsafe code is required to actually access it.
+
+To ensure compatibility, the `#[repr(type = ...)]` form is required if the type is not one of the known primitive types. Note that this form is not necessarily equivalent to using the primitive representations directly, since shadowing is possible; for example, if you did `type u32 = u8` and then `#[repr(type = u32)]`, this would be equivalent to `#[repr(u8)]`, not `#[repr(u32)]`.
+
+You can use any type alias in the `repr` attribute, but it *must* be an alias to an accepted primitive type like `u8` or `i32`, and cannot be a pointer, reference, struct, etc.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+The `repr` attribute now accepts a `type = ...` argument to indicate a resolved path instead of a well-known primitive type. If those the path resolves to a type alias to a valid primitive type which can be used in the `repr` attribute, that will be used as the actual discriminant representation.
+
+An additional, automatically-applicable lint should be added that warns a user if they use `type = ...` for a well-known primitive type, since adding `type = ` instead of using the type directly introduces the possibility of shadowing. (For example, `#[repr(type = u32)]` becomes `#[repr(u32)]`.)
+
+Similarly, an automatically-applicable lint should be added that warns a user if a `repr` argument references in an-scope type alias without the `type = ` prefix. (For example, `#[repr(MyType)]` becomes `#[repr(type = MyType)]`.)
+
+# Drawbacks
```

Comment:
1. `#[derive(FromBytes)]` could conservatively reject `#[repr(type = T)]`
2. furthermore the code could simply *not* do the check directly in the proc macro, but generate a const block

    ```rust
    const { assert!(
        size_of::<X>() < size_of::<usize>() &&
        (1usize << (8 * size_of::<X>())) == variant_count::<X>()
    ) };
    // the unstable `variant_count::<X>()` can be replaced 
    // by the proc-macro-computed integer literal
    ```

    to let rustc to do the check.

Reply:
So, you should be able to do this with one of the pre-existing associated constants for these types-- for example, `BITS`. It might require moving some of the logic from erroring directly in the macro to erroring in a constant, for example, but it still feels doable.

Reply:
Agreed that this isn't insurmountable — just a complication. I believe `zerocopy` will need to conservatively reject `#[repr(type = T)]`, since the second option will not be available until `variant_count` is stabilized.

Reply:
@jswrenn a proc macro like `#[derive(FromBytes)]` in particular does *not* need a stable variant_count since it already has the precise variant_count from https://docs.rs/syn/latest/syn/struct.DataEnum.html (`enm.variants.len()`).

Reply:
Ah, I misunderstood. `1usize << (8 * size_of::<X>())` would do the trick, assuming we can move the check into a const context.

Reply:
Yeah, I definitely think it's worth listing as a drawback, but as mentioned, I think that the benefits of having this outweigh the concerns of having to move these kinds of validations to a const context. Especially since the users of a library like `zerocopy` (hopefully) understand more about the intricacies of what they're doing to be able to parse the resulting error.

Reply:
Going to mark this as resolved since I added a note about this to the RFC summarising this discussion.